### ISSUE TITLE
Comma delimiter shouldn't be used for positional arguments

### ISSUE THREAD

**Issue by equeim** [jarro2783/cxxopts#228]
If you use cxxopts like this:
```
std::vector<std::string> files;
cxxopts::Options opts(appName, versionString);
opts.add_options()
    ("files", "", cxxopts::value<std::vector<std::string>>(files));
opts.parse_positional("files");
```
And call executable like this (notice the quotes):
`./foo /home/bar/file_without_comma "/home/bar/file, with comma"`
files vector will contain three items instead of two, despite presence of quotes.

Of course you can workaround this behaviour, either by defining CXXOPTS_VECTOR_DELIMITER with null character which will also prevent you from using option arguments with delimiters (which is not desirable), or by iterating over ParseResult::arguments() which in this case contains two items with key "files". I, however, feel that this behaviour is unintuitive.

**Comment by jarro2783** [comment]
Is having commas in your filenames common? The comma is only the default, you can redefine it yourself. I think removing this functionality is also not desirable because people now rely on it.

I might have to look at what other parsers do with this, since a few people were requesting a way to parse a list of arguments by some delimiter at some point, which is why this feature was added.

**Comment by equeim** [comment]
It not question of how common this is (although it is quite common, yes), if a program takes filenames as arguments it must process them correctly. On Unix-like systems filenames can contain any characters except '\0', even newlines. Besides, positional arguments are already lists by their very nature, there is simply no need to split them by delimiters. Although I of course understand backwards compatibility concerns.

The problem with redefining CXXOPTS_VECTOR_DELIMITER is that it used to parse all std::vector arguments. What if some cxxopts user wanted to to parse regular list argument, like `--foo=42,666` splitted by commas, _and_  to parse positional arguments without special handling for commas in the same program? They would have to either redefine CXXOPTS_VECTOR_DELIMITER and do splitting themselves, or leave CXXOPTS_VECTOR_DELIMITER untouched and iterate over ParseResult::arguments() to get unsplitted positional arguments.

I think the best solution would be to be able to either configure list delimiter per option or simply disable delimiter splitting for select option. I have no idea how hard it would be implement though.

**Comment by jarro2783** [comment]
Yeah that's a good point. Doing this per argument might be more reasonable.

This isn't just for positional arguments though. It's possible to have some non-positional arguments that can appear multiple times. So instead of writing `--arg a --arg b --arg c` you can then just write `--arg a,b,c`.

**Comment by mwestphal** [comment]
Actual usecase in gvfs, in  f3d-app/f3d#1044, where a gvfs user uses `,` in the file path for perfectly valid reasons.

**Comment by lazysegtree** [comment]
@mwestphal @bdon @KrystalDelusion @equeim 
 I see that latest cxxopts doesn't have that behaviour. For positional args, its just maps every argv to one arg. 

It was fixed in 
- https://github.com/jarro2783/cxxopts/pull/398

```
➜  ~/Workspace/kuknitin/cxxopts git:(temp-shortarg-fix) ✗ [12:17:38] g++ example/font_maker.cpp && ./a.out --name "NotoSans" "./tmp" "./NotoSans[wdth,wght].ttf"
output_dir = ./tmp
fonts.size() = 1 | fonts = [./NotoSans[wdth,wght].ttf]
➜  ~/Workspace/kuknitin/cxxopts git:(temp-shortarg-fix) ✗ [12:17:43] g++ example/positional.cpp && ./a.out --vo 1,2,3 /home/bar/file_without_comma "/home/bar/file, with comma"
val.size() = 2 | val = [/home/bar/file_without_comma, /home/bar/file, with comma]
val.size() = 3 | val = [1, 2, 3]
➜  ~/Workspace/kuknitin/cxxopts git:(temp-shortarg-fix) ✗ [12:17:52]
```

Programs

<details><summary>Details</summary>
<p>

```c++

#include <iostream>
#include "debug.h"

#include "../include/cxxopts.hpp"

using namespace std;

int main(int argc, char* argv[])
{
     cxxopts::Options cmd_options("font-maker", "Create font PBFs from TTFs or OTFs.");
    cmd_options.add_options()
        ("output", "Output directory (to be created, must not already exist)", cxxopts::value<string>())
        ("fonts", "Input font(s) (as TTF or OTF)", cxxopts::value<vector<string>>())
        ("name", "Override output fontstack name", cxxopts::value<string>())
        ("help", "Print usage")
    ;
    cmd_options.positional_help("<OUTPUT_DIR> <INPUT_FONT> [INPUT_FONT2 ...]");
    cmd_options.parse_positional({"output","fonts"});

    auto result = cmd_options.parse(argc, argv);
    if (result.count("help"))
    {
      cout << cmd_options.help() << endl;
      exit(0);
    }
    if (result.count("output") == 0 || result.count("fonts") == 0) {
        cout << cmd_options.help() << endl;
        exit(1);
    }
    auto output_dir = result["output"].as<string>();
    auto fonts = result["fonts"].as<vector<string>>();
    DEBUG(output_dir);
    DEBUG(fonts.size(), fonts);
}
```

```c++

#include <iostream>
#include "debug.h"

#include "../include/cxxopts.hpp"

int main(int argc, const char* argv[])
{
  cxxopts::Options options("TestProgram", "Sample argument parsing");
  options.add_options()
    ("vo", "Vector", cxxopts::value<std::vector<std::string>>())
    ("o,option", "Opt", cxxopts::value<std::vector<std::string>>())
    ;
  options.parse_positional({"o"});
  auto result = options.parse(argc, argv);

  if(result.count("option")) {
    auto val = result["option"].as<std::vector<std::string>>();
    DEBUG(val.size(), val);
  }
  if(result.count("vo")) {
    auto val = result["vo"].as<std::vector<std::string>>();
    DEBUG(val.size(), val);
  }
}
```

</p>
</details>

**Comment by KrystalDelusion** [comment]
@lazysegtree I can try to investigate this again, but YosysHQ/yosys#5176 has this to say

> Say you want to to call `yosys test.v -p prep -p 'eval -table a,b'`.  The arguments passed to main are
```
yosys
test.v
-p
prep
-p
eval -table a,b
```
> The resulting `result["p"]` is
```
prep
eval -table a
b
```

We are using v.3.2.1 (4bf61f08) and have been since we started using the library, which includes the PR you mentioned.  Is it possible something else has changed since 3.2.1, or that perhaps this is platform dependent (my investigation was using Ubuntu 20.04 and clang)?

**Comment by lazysegtree** [comment]
@KrystalDelusion What you are showing is still as I expect the program to behave. Not splitting by delimiter was only implemented for positional arguments. In your example, `-p` is a non positional argument.

I can also reproduce the same in v3.3.1

```bash
➜  ~/Workspace/kuknitin/cxxopts git:(temp-shortarg-fix) ✗ [8:17:02] g++ example/positional.cpp && ./a.out --vo 1 --vo 2 --vo 3,4 /home/bar/file_without_comma "/home/bar/file, with comma"
val.size() = 2 | val = [/home/bar/file_without_comma, /home/bar/file, with comma]
val.size() = 4 | val = [1, 2, 3, 4]
➜  ~/Workspace/kuknitin/cxxopts git:(temp-shortarg-fix) ✗ [8:20:07]
```

**Comment by KrystalDelusion** [comment]
My issue may be a separate one then, because my expectation is that `a.out --vo 1 --vo 2,3 --vo "4,5"` would produce 4 values for `vo`: `[1, 2, 3, "4,5"]`

**Comment by lazysegtree** [comment]
```
a.out --vo 1 --vo 2,3 --vo "4,5" would produce 4 values for vo: [1, 2, 3, "4,5"]
```

@KrystalDelusion This is actually not even possible as shell will unquote `"4,5"` . 
This is what program gets

```bash
➜  ~/Workspace/kuknitin/cxxopts git:(temp-shortarg-fix) ✗ [8:20:07] g++ example/positional.cpp && ./a.out --vo 1 --vo 2,3 --vo "4,5"
i = 0 | argv[i] = ./a.out
i = 1 | argv[i] = --vo
i = 2 | argv[i] = 1
i = 3 | argv[i] = --vo
i = 4 | argv[i] = 2,3
i = 5 | argv[i] = --vo
i = 6 | argv[i] = 4,5
val.size() = 5 | val = [1, 2, 3, 4, 5]
➜  ~/Workspace/kuknitin/cxxopts git:(temp-shortarg-fix) ✗ [8:32:33]
```

**Comment by KrystalDelusion** [comment]
Yes I know, which is why our workaround was to disable the delimiter entirely even at the cost of `--vo 1,2` type arguments.

**Comment by lazysegtree** [comment]
@KrystalDelusion So a feature of disabling delimiter per argument should help here.

Something like `cxxopts::value<std::vector<std::string>>()->disable_vector_delimit()`

I am not sure how easy it is. I am going to look around the code and see.

**Comment by jarro2783** [comment]
It looks like the original issue is solved, I'll close this. Feel free to open a new one if there is another problem.
